src/user-event/press/press.ts

```diff
@@ -1,16 +1,118 @@
+  duration: number;
+};
 
 export async function press(
   this: UserEventInstance,
   element: ReactTestInstance
-) {
-  // TODO provide real implementation
-  dispatchHostEvent(element, 'pressIn', EventBuilder.Common.touch());
+): Promise<void> {
+  await basePress(this.config, element);
+}
 
-  await wait(this.config);
-  dispatchHostEvent(element, 'press', EventBuilder.Common.touch());
-  dispatchHostEvent(element, 'pressOut', EventBuilder.Common.touch());
+export async function longPress(
+  this: UserEventInstance,
+  element: ReactTestInstance,
+  options: PressOptions = { duration: 500 }
+): Promise<void> {
+  await basePress(this.config, element, options);
 }
+
+const basePress = async (
+  config: UserEventInstance['config'],
+  element: ReactTestInstance,
+  options: PressOptions = { duration: 0 }
+): Promise<void> => {
+  // Text and TextInput components are mocked in React Native preset so the mock
+  // doesn't implement the pressability class
+  // Thus we need to call the props directly on the host component
+  const isEnabledHostText =
+    filterNodeByType(element, getHostComponentNames().text) &&
+    isPointerEventEnabled(element) &&
+    !element.props.disabled;
+  const isEnabledTextInput =
+    filterNodeByType(element, getHostComponentNames().textInput) &&
+    isPointerEventEnabled(element) &&
+    element.props.editable !== false;
+
+  if (isEnabledHostText || isEnabledTextInput) {
+    const { onPressIn, onPress, onPressOut } = element.props;
+    await wait(config);
+    if (onPressIn) {
+      onPressIn(EventBuilder.Common.touch());
+    }
+    if (onPress) {
+      onPress(EventBuilder.Common.touch());
+    }
+    await wait(config, options.duration);
+    if (onPressOut) {
+      if (DEFAULT_MIN_PRESS_DURATION - options.duration > 0) {
+        await wait(config, DEFAULT_MIN_PRESS_DURATION - options.duration);
+      }
+      onPressOut(EventBuilder.Common.touch());
+    }
+  }
+
+  if (isEnabledTouchResponder(element)) {
+    await triggerPressEvent(config, element, options);
+    return;
+  }
+
+  const hostParentElement = getHostParent(element);
+  if (!hostParentElement) {
+    return;
+  }
+
+  await basePress(config, hostParentElement, options);
+};
+
+const triggerPressEvent = async (
+  config: UserEventInstance['config'],
+  element: ReactTestInstance,
+  options: PressOptions = { duration: 0 }
+) => {
+  const areFakeTimersEnabled = jestFakeTimersAreEnabled();
+  if (!areFakeTimersEnabled) {
+    warnAboutRealTimers();
+  }
+
+  await wait(config);
+
+  await act(async () => {
+    element.props.onResponderGrant({
+      ...EventBuilder.Common.touch(),
+      dispatchConfig: { registrationName: 'onResponderGrant' },
+    });
+
+    await wait(config, options.duration);
+
+    element.props.onResponderRelease({
+      ...EventBuilder.Common.touch(),
+      dispatchConfig: { registrationName: 'onResponderRelease' },
+    });
+
+    if (DEFAULT_MIN_PRESS_DURATION - options.duration > 0) {
```

Comment:
[mid] q: why do we need to do additional waiting to reach at least min press duration limit here and in the Text/TextInput block above?

I've found [this line](https://github.com/facebook/react-native/blob/06668fcbacd750771f1d53cce829dc55e86f3f3c/packages/react-native/Libraries/Pressability/Pressability.js#L955) in RN source. This seems to be an un-documented internal implementation detail of RN. I wonder if:
1) If it's already being applied so that we would not need to add manual wait
2) If not is it worth relying on such implementation details.

@pierrezimmermannbam wdyt?

Reply:
What happens in RN code when the press is finished is:

- the delay for the pressOut is computed: it's equal to `minimum press duration` -  `press duration`(see [source code](https://github.com/facebook/react-native/blob/4c944540f732c6055d447ecaf37d5c8f3eec1bc4/packages/react-native/Libraries/Pressability/Pressability.js#L810)
- if the delay is > 0 then pressOut is programmed through a setTimeout with the delay computed at the above step (see [source code](https://github.com/facebook/react-native/blob/4c944540f732c6055d447ecaf37d5c8f3eec1bc4/packages/react-native/Libraries/Pressability/Pressability.js#L814)) 

In the implementation for `Pressable`, we wait because if the press duration is shorter than the minimum press duration, then `onPressOut`has not been called yet. This is problematic for two reasons: 
- users would expect `onPressOut` to have been called
- it results in an act warning

This is why we wait in this case and in the second scenario, we could just call `onPressOut` without waiting, it would just be less realistic. The upside of not waiting is that tests that don't use fake timers would be faster

Reply:
Could you confirm that I get that right: the `Pressable` is actually delaying the action, so that we wait at least 130 ms, event when running under tests. If we do not wait to that 130 ms, then the event handler will be called outside of async `act` causing async `act` warning? So we are actually waiting to prevent async `act` warning from happening.

Reply:
Yes that's correct. We're waiting both to prevent act warning but also to have `onPressOut` called synchronously so user doesn't have to wait for it